### PR TITLE
Allow to hide settings in preprocessed configuration file

### LINKED ISSUE #52207: Preprocessed config file with hidden/in-memory parts
https://github.com/ClickHouse/ClickHouse/issues/48291 and https://github.com/ClickHouse/ClickHouse/pull/50986 try to avoid plaintext passwords in configuration files. A second angle of attack is the preprocessed XML file which might contains sensitive data (fused from multiple XML/JSON files with varying access rights). We should check if parts of the configuration tree can be marked "hidden" such they are not written into the preprocessed XML file in the first place.

(perhaps it is easier to avoid writing a preprocessed XML file altogether, i.e. rather do both steps of preprocessing and populating the server config from the preprocessed data entirely in-memory)

### PR DESCRIPTION
**Pull request by rvasin**
Closes #52207

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added the "hide_in_preprocessed" attribute to ClickHouse's server configuration XML dialect. This is a mechanism to hide certain settings from appearing in preprocessed server configuration files. Useful e.g. for passwords or private keys that should not appear verbatim in files.

### CODE REVIEW

**Comment by rvasin** [comment]
@rschu1ze could you please be reviewer of this PR and assign `can be tested` label?

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit 7a85c9af5f3d85d53267fb8c66b9b9eb751b77c7 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53818/7a85c9af5f3d85d53267fb8c66b9b9eb751b77c7/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by rvasin** [comment]
The failed test `02494_optimize_group_by_function_keys_and_alias_columns` in Fasttest is not related. It seems flaky. I will try to update by branch with master to see the next run of Fasttest.

File: docs/en/operations/configuration-files.md [hunk 1]
```diff
@@ -102,6 +102,23 @@ Example:
 961F000000040000000000EEDDEF4F453CFE6457C4234BD7C09258BD651D85
 ```
 
+## Hiding Configuration {#hidden}
```
Comment (rschu1ze) [inline docs/en/operations/configuration-files.md:105]: I think this section conceptually belongs into the previous section about encryption. What about
- renaming the caption at l. 68 to "Encrypting and Hiding Configuration"
- explaining in l. 104 that even with encryption, the configuration file after preprocessing will contain the cleartext element. In case this is a problem, we suggest two alternatives: either set file permissions of the preprocessed file to 600 or use the hidden attribute.
- then explain what the hidden attribute does using the same example as for encryption (you kind of do that already)

File: tests/integration/test_config_hidden_attributes/configs/config.xml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<clickhouse>
+    <max_table_size_to_drop hidden="true">60000000000</max_table_size_to_drop>
```
Comment (rschu1ze) [inline tests/integration/test_config_hidden_attributes/configs/config.xml:2]: Thinking about naming. "hidden" as attribute name is not as intuitive as existing attributes "from_env" or "from_zk" or "encrypted_by". What about "hide_in_preprocessed" instead?

File: tests/integration/test_config_hidden_attributes/configs/config.xml [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    <max_partition_size_to_drop hidden="false">40000000000</max_partition_size_to_drop>
```
Comment (rschu1ze) [inline tests/integration/test_config_hidden_attributes/configs/config.xml:3]: The code uses `Poco::NumberParser::parseBool(element.getAttribute("hidden"))`. It would be nice to validate that `hidden = "1"` achieves the same result (just to avoid surprises for people who prefer "1" over "true").

File: src/Common/Config/ConfigProcessor.cpp [hunk 4]
```diff
@@ -792,6 +812,21 @@ void ConfigProcessor::decryptEncryptedElements(LoadedConfig & loaded_config)
 
 #endif
 
+XMLDocumentPtr ConfigProcessor::hideElements(LoadedConfig & loaded_config)
+{
+    XMLDocumentPtr new_preprocessed_xml = new Poco::XML::Document;
+
+    for (Node * node = loaded_config.preprocessed_xml->firstChild(); node; node = node->nextSibling())
```
Comment (rschu1ze) [inline src/Common/Config/ConfigProcessor.cpp:819]: So l. 817-825 copy of the preprocessed XML structucture, right? Why is that necessary, can't we call `hideRecursive()` directly on the existing preprocessed XML?

File: src/Common/Config/ConfigProcessor.cpp [hunk 5]
```diff
@@ -792,6 +812,21 @@ void ConfigProcessor::decryptEncryptedElements(LoadedConfig & loaded_config)
 
 #endif
 
+XMLDocumentPtr ConfigProcessor::hideElements(LoadedConfig & loaded_config)
+{
+    XMLDocumentPtr new_preprocessed_xml = new Poco::XML::Document;
```
Comment (rschu1ze) [inline src/Common/Config/ConfigProcessor.cpp:817]: `hideElements` is currently called with the preprocessed config but it must not be necessarily be so. We should call `new_preprocessed_xml` rather `xml_tree_copy`.

File: src/Common/Config/ConfigProcessor.cpp [hunk 6]
```diff
@@ -840,7 +875,7 @@ void ConfigProcessor::savePreprocessedConfig(LoadedConfig & loaded_config, std::
         writer.setNewLine("\n");
         writer.setIndent("    ");
         writer.setOptions(Poco::XML::XMLWriter::PRETTY_PRINT);
-        writer.writeNode(preprocessed_path, loaded_config.preprocessed_xml);
+        writer.writeNode(preprocessed_path, hideElements(loaded_config));
```
Comment (rschu1ze) [inline src/Common/Config/ConfigProcessor.cpp:878]: Let's make two lines out of one:
```
XMLDocumentPtr preprocessed_xml_without_hidden_elements = hideElements(loadedConfig.preprocessed_xml);
writer.writeNode(preprocessed_path, preprocessed_xml_without_hidden_elements);
```

File: src/Common/Config/ConfigProcessor.cpp (hunk 4 above)
Comment (rvasin) [reply src/Common/Config/ConfigProcessor.cpp:819]: In the first commits I applied `hideRecursive()` directly on loaded_config.preprocessed_xml. But then I discovered that loaded_config.configuration in fact is a pointer to `loaded_config.preprocessed_xml`. So cutting nodes from `preprocessed_xml` also influences on `configuration` (missing values will be reset to defaults). That why I:
1. create copy of  `preprocessed_xml` 
2. hide nodes on copy.
3. save copy as preprocessed config file.

Comment (rschu1ze) [reply src/Common/Config/ConfigProcessor.cpp:819]: Okay, but please add a short comment to the code then.

File: tests/integration/test_config_hidden_attributes/configs/config.xml (hunk 2 above)
Comment (rvasin) [reply tests/integration/test_config_hidden_attributes/configs/config.xml]: Done. Renamed.

File: tests/integration/test_config_hidden_attributes/configs/config.xml (hunk 3 above)
Comment (rvasin) [reply tests/integration/test_config_hidden_attributes/configs/config.xml]: Done. Added test for "1" value.

File: src/Common/Config/ConfigProcessor.cpp (hunk 4 above)
Comment (rvasin) [reply src/Common/Config/ConfigProcessor.cpp]: Done. Added comment. Again. Please text a look at the comment's text and it's position. Maybe we may improve text or put it into more clear position.

File: src/Common/Config/ConfigProcessor.cpp (hunk 5 above)
Comment (rvasin) [reply src/Common/Config/ConfigProcessor.cpp]: Done

File: src/Common/Config/ConfigProcessor.cpp (hunk 6 above)
Comment (rvasin) [reply src/Common/Config/ConfigProcessor.cpp]: Done

**Review by rschu1ze — APPROVED** [review]

**Comment by rvasin** [comment]
I am working on fixing the integration tests. It when ClickHouse is built with ASAN then in fails to restart ClickHouse in integration test using `node.restart_clickhouse()`

For example, 
```python
def test_hide_in_preprocessed(started_cluster):
    assert (
        node.query(
            "select value from system.server_settings where name ='max_thread_pool_free_size'"
        )
        == "2000\n"
    )
    node.restart_clickhouse()
    assert (
        node.query(
            "select value from system.server_settings where name ='max_thread_pool_free_size'"
        )
        == "2000\n"
    )
```
It will display "Exception: Cannot start ClickHouse, see additional info in logs"  which is produced by `elf.start_clickhouse(stop_start_wait_sec)`.

It's not clear how is it related with by code and with ASAN. But surely it's related. I am going to fix it.

**Comment by rvasin** [comment]
Thanks to ASAN build, probably the failed integration tests were fixed by changing:
```c++
Node * new_node = xml_tree_copy->importNode(node, true);
```

to
```c++
NodePtr new_node = xml_tree_copy->importNode(node, true);
```

Because NodePtr is AutoPtr:
base/poco/Foundation/include/Poco/AutoPtr.h
```c++
class AutoPtr
/// AutoPtr is a "smart" pointer for classes implementing
/// reference counting based garbage collection.
```

**Comment by rvasin** [comment]
@rschu1ze "All checks have passed".
